### PR TITLE
 utility: introduce pipeline library

### PR DESCRIPTION
**Pull request by neverchanje**
This PR includes:

**pipeline**
see the comments and unit tests to know how to use it. see this ppt [pipeline.pdf](https://github.com/XiaoMi/rdsn/files/1927810/pipeline.pdf) that provides the introduction of how it was designed.

**[abseil](https://github.com/abseil/abseil-cpp)'s implementation of std::apply**
- include/dsn/utility/absl/base/internal/invoke.h
- include/dsn/utility/absl/utility/utility.h
-  include/dsn/utility/apply.h
apply allows a function of any type to be called with arguments wrapped in `std::tuple`. It's crucial for implementation of pipeline.

**pipeline_test**
unit test
we still need more.

### CODE REVIEW

**Comment by neverchanje** [comment]
可以的，整个代码尽可能保证 zero-overhead，所以可以多从用户视角看一下整个接口设计，代码也不多，真的遇到问题要细抠的时候也不会很麻烦。

File: src/core/tests/pipeline_test.cpp [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+{
+    clientlet tracker;
+
+    struct stage2 : pipeline::when<>, pipeline::result<>
+    {
+        void run() override { step_down_next_stage(); }
+    };
+
+    {
+
```
Comment (qinzuoyan) [inline src/core/tests/pipeline_test.cpp]: useless new line

File: include/dsn/utility/apply.h [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+#pragma once
+
+#include <dsn/utility/absl/utility/utility.h>
+
+namespace dsn {
+
+using ::absl::apply;
+
+} // namespace dsn
```
Comment (qinzuoyan) [inline include/dsn/utility/apply.h]: what character at last of the line?

File: include/dsn/utility/apply.h [hunk 3]
```diff
@@ -0,0 +1,35 @@
+/*
```
Comment (qinzuoyan) [inline include/dsn/utility/apply.h]: incorrect licence

File: src/core/tests/pipeline_test.cpp [hunk 4]
```diff
@@ -0,0 +1,89 @@
+/*
```
Comment (qinzuoyan) [inline src/core/tests/pipeline_test.cpp]: incorrect licence

File: include/dsn/cpp/pipeline.h [hunk 5]
```diff
@@ -0,0 +1,294 @@
+/*
```
Comment (qinzuoyan) [inline include/dsn/cpp/pipeline.h]: incorrect licence

**Pull request merged**